test/jdk/java/net/httpclient/UserAuthWithAuthenticator.java

```diff
@@ -37,17 +49,122 @@
+                 OutputStream os = t.getResponseBody()) {
+                byte[] bytes = is.readAllBytes();
+                authValue = t.getRequestHeaders().firstValue("Authorization").orElse("");
+                t.sendResponseHeaders(200, response.length());
+                os.write(response.getBytes(US_ASCII));
+                t.close();
+            }
+        }
+
+        String authValue() {return authValue;}
+    }
+
+    // if useHeader is true, we expect the Authenticator was not called
+    // and the user set header used. If false, Authenticator must
+    // be called and the user set header not used.
+
+    static void h2Test(final boolean useHeader) throws Exception {
+        SSLContext ctx;
+        HttpTestServer h2s = null;
+        HttpClient client = null;
+        ExecutorService ex=null;
+        try {
+            ctx = new SimpleSSLContext().get();
+            ex = Executors.newCachedThreadPool();
+            InetAddress addr = InetAddress.getLoopbackAddress();
+
+            h2s = HttpTestServer.of(new Http2TestServer(addr, "::1", true, 0, ex,
+                    10, null, ctx, false));
+            AuthTestHandler h = new AuthTestHandler();
+            var context = h2s.addHandler(h, "/test1");
+            context.setAuthenticator(new BasicAuthenticator("realm") {
+                public boolean checkCredentials(String username, String password) {
+                    if (useHeader) {
+                        return username.equals("user") && password.equals("pwd");
+                    } else {
+                        return username.equals("serverUser") && password.equals("serverPwd");
+                    }
+                }
+            });
+            h2s.start();
+
+            int port = h2s.getAddress().getPort();
+            ServerAuth sa = new ServerAuth();
+            var plainCreds = "user:pwd";
+            var encoded = java.util.Base64.getEncoder().encodeToString(plainCreds.getBytes(US_ASCII));
+
+            URI uri = URIBuilder.newBuilder()
+                 .scheme("https")
+                 .host(addr.getHostAddress())
+                 .port(port)
+                 .path("/test1/foo.txt")
+                 .build();
+
+            HttpClient.Builder builder = HttpClient.newBuilder()
+                    .sslContext(ctx)
+                    .executor(ex);
+
+            //if (!useHeader) {
+                builder.authenticator(sa);
+            //}
+            client = builder.build();
+
+            HttpRequest req = HttpRequest.newBuilder(uri)
+                    .version(HttpClient.Version.HTTP_2)
+                    .header(useHeader ? "Authorization" : "X-Ignore", encoded)
+                    .GET()
+                    .build();
+
+            HttpResponse<String> resp = client.send(req, HttpResponse.BodyHandlers.ofString());
+            if (useHeader) {
+                assertTrue(h.authValue() == null, "Expected user set header to be set");
```

Comment:
authValue is null if the handler was never called; is that what we want to check here?

Reply:
So, initially I thought the problem here was that the request needed to be retried and when using a manually set Authorization header, retries do not happen automatically (Question: should this be documented in the implNote? I think it probably should be).

Then I realised that was **not** the problem because the Authorization is being provided pre-emptively and no retry is necessary. The actual issue was that the manually set Authorization header was malformed and was missing the "Basic" scheme prefix.